pkg/controller/risingwave_controller.go

```diff
@@ -241,9 +256,40 @@ func (c *RisingWaveController) reactiveWorkflow(risingwaveManger *object.RisingW
 		return ctrlkit.Continue()
 	})
 	syncConfigs := mgr.SyncConfigConfigMap()
-	syncMetaComponent := ctrlkit.ParallelJoin(mgr.SyncMetaService(), mgr.SyncMetaDeployments())
+
+	cloneSetCRDsInstalledBarrier := mgr.NewAction(RisingWaveAction_BarrierCloneSetCRDsInstalled, func(ctx context.Context, l logr.Logger) (ctrl.Result, error) {
+		crd, err := utils.GetCustomResourceDefinition(c.Client, ctx, metav1.GroupKind{
+			Group: "apps.kruise.io",
+			Kind:  "CloneSet",
+		})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return ctrlkit.Exit()
+			}
+			return ctrlkit.RequeueIfErrorAndWrap("unable to find CRD for Cloneset", err)
+		}
+		return ctrlkit.ExitIf(!utils.IsVersionServingInCustomResourceDefinition(crd, "v1alpha1"))
+	})
+
+	advancedStsCRDsInstalledBarrier := mgr.NewAction(RisingWaveAction_BarrierAdvancedStsCRDsInstalled, func(ctx context.Context, l logr.Logger) (ctrl.Result, error) {
+		crd, err := utils.GetCustomResourceDefinition(c.Client, ctx, metav1.GroupKind{
+			Group: "apps.kruise.io",
+			Kind:  "StatefulSet",
+		})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return ctrlkit.Exit()
+			}
+			return ctrlkit.RequeueIfErrorAndWrap("unable to find CRD for Advanced Statefulset", err)
+		}
+
+		return ctrlkit.ExitIf(!utils.IsVersionServingInCustomResourceDefinition(crd, "v1beta1"))
+	})
+
+	syncMetaComponent := ctrlkit.ParallelJoin(mgr.SyncMetaService(), mgr.SyncMetaDeployments(), ctrlkit.If(c.openKruiseAvailable, ctrlkit.SequentialJoin(cloneSetCRDsInstalledBarrier, mgr.SyncMetaCloneSets())))
```

Comment:
Shall we break it into lines to make it easier to read?

Reply:
have been fixed in the latest commit.